src/app/components/chat/jh-chat.directive.js

```diff
@@ -25,18 +27,95 @@
     };
 
     function jhChatLink(scope) {
-      scope.$watch('vm.messages.length', function(newVal) {
-        /* Update messages */
-        if (newVal !== undefined) {
-          // Scroll to bottom of messages list.
-          var messagesList = document.getElementById('jh-chat-messages');
-          setTimeout(function() {
-            messagesList.scrollTop = messagesList.scrollHeight;
-          }, 100);
+      scope.entriesCount = 0;       // number of all types of entries
+      scope.messagesCount = 0;      // number of entries of type 'chatMsg'
+      scope.lastSeenMessage = 0;    // last seen entry of type 'chatMsg'
+      scope.isChatVisible = false;
+
+      scope.$watch('vm.logEntries().length', function() {
+        // Scroll to bottom of messages list.
+        var messagesList = document.getElementById('chat-messages-box');
+        setTimeout(function() {
+          messagesList.scrollTop = messagesList.scrollHeight;
+        }, 100);
+      });
+
+      scope.$watch('vm.logEntries()', function(entries) {
+        // True only at the beginning when there are no chat messages
+        if (entries.length === 0) {
+          return;
+        }
+
+        // Update the chat messages count
+        for (var i = scope.entriesCount; i < entries.length; i++) {
+          if (entries[i].type === 'chatMsg') {
+            scope.messagesCount++;
+            if (scope.isChatVisible) {
+              scope.lastSeenMessage++;
+            } else {
+              var chatHeader = document.getElementById('chat-header');
+              chatHeader.innerHTML = scope.messagesCount - scope.lastSeenMessage;
+            }
+          } else {
+            // We don't notify the user when the entry is not of type 'chatMsg'
+          }
+        }
+        scope.entriesCount = entries.length;
+      });
+
+      scope.$watch('vm.isChatVisible', function(isVisible) {
+        if (isVisible === undefined) {
+          return;
+        }
+        var chatHeader = document.getElementById('chat-header');
+        var unreadMessages = scope.messagesCount - scope.lastSeenMessage;
+        if (isVisible) {
+          scope.isChatVisible = true;
+          scope.lastSeenMessage = scope.messagesCount;
+          chatHeader.innerHTML = ""; // Display nothing when chat is open
+          if (unreadMessages === 0) {
+            $("#chat-header").toggleClass("read");
+          } else {
+            $("#chat-header").toggleClass("unread");
+          }
+        } else {
+          scope.isChatVisible = false;
+          if (unreadMessages === 0) {
+            chatHeader.innerHTML = "Join the chat!";
+            $("#chat-header").toggleClass("read");
```

Comment:
Basically you are calling `toggleClass` in several places. What about setting some `cssClass` variable and just call `toggleClass` at the end? IMHO it would be a cleaner approach.